### PR TITLE
Refactor/sdk preview

### PR DESCRIPTION
**Pull request by L-jasmine**
Refactor the SDK to resolve memory-unsafe bugs caused by its usage.

### CODE REVIEW

**Comment by juntao** [comment]
Hello, I am a [code review bot](https://github.com/flows-network/github-pr-summary/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

**Comment by juntao** [comment]
@apepkuss @CaptainVincent Please review. We should aim to merge soon. Thanks.

**Comment by apepkuss** [comment]
@L-jasmine Please consider to merge both [standalone](https://github.com/WasmEdge/wasmedge-rust-sdk/blob/main/.github/workflows/standalone.yml) and [rust-static-lib](https://github.com/WasmEdge/wasmedge-rust-sdk/blob/main/.github/workflows/rust-static-lib.yml) workflows into this PR. These two workflows are necessary for quality assurance. For example, both `runwasi` and `Docker` teams have dependencies on these two modes.

**Comment by apepkuss** [comment]
@DarumaDocker Does this PR satisfy your requirement of using the `wasi-nn_ggml` plugin? In the current release (v.0.13.2), it's supported by [PluginManager::nn_preload](https://wasmedge.github.io/wasmedge-rust-sdk/wasmedge_sdk/plugin/struct.PluginManager.html#method.nn_preload) API.

**Review by apepkuss — CHANGES_REQUESTED** [review]
Please remove them if they are necessary anymore. Thanks!

**Review by apepkuss — COMMENTED** [review]
why skip Vmbuilder-related tests here?

File: .github/workflows/ci-build.yml [hunk 1]
```diff
@@ -159,7 +159,7 @@ jobs:
           export WASMEDGE_BUILD_DIR="$(pwd)/WasmEdge/build"
           export WASMEDGE_PLUGIN_PATH="$(pwd)/WasmEdge/build/plugins/wasmedge_process"
           export LD_LIBRARY_PATH="$(pwd)/WasmEdge/build/lib/api"
-          cargo test --workspace --locked --features aot,async,wasmedge_process,ffi -- --nocapture --test-threads=1
+          cargo test --workspace --locked --features aot,async,wasmedge_process,ffi -- --nocapture --test-threads=1 --skip test_vmbuilder
```
Comment (apepkuss) [inline .github/workflows/ci-build.yml:162]: Why skip VmBuilder-related tests here?

**Comment by apepkuss** [comment]
@L-jasmine Please refer to #91 to update the workspace `Cargo.toml`.

File: crates/wasmedge-sys/src/plugin.rs [hunk 2]
```diff
@@ -194,21 +197,20 @@ impl Plugin {
...
     }
 
-    /// Provides a raw pointer to the inner Plugin context.
-    #[cfg(feature = "ffi")]
-    #[cfg_attr(docsrs, doc(cfg(feature = "ffi")))]
-    pub fn as_ptr(&self) -> *const ffi::WasmEdge_PluginContext {
+    /// # Safety
+    ///
+    /// The lifetime of the returned pointer must not exceed that of the object itself.
+    pub unsafe fn as_ptr(&self) -> *const ffi::WasmEdge_PluginContext {
```
Comment (apepkuss) [inline crates/wasmedge-sys/src/plugin.rs:208]: No need to use `unsafe` here as only the return value is of pointer type.

File: src/io.rs [hunk 3]
```diff
@@ -38,7 +38,7 @@ impl_wasm_val_type!(i64, ValType::I64);
 impl_wasm_val_type!(f32, ValType::F32);
 impl_wasm_val_type!(f64, ValType::F64);
 impl_wasm_val_type!(i128, ValType::V128);
-impl_wasm_val_type!(FuncRef, ValType::FuncRef);
+// impl_wasm_val_type!(FuncRef, ValType::FuncRef);
```
Comment (apepkuss) [inline src/io.rs:41]: Please remove it if it's not useful. Thanks.

File: src/io.rs [hunk 4]
```diff
@@ -57,19 +57,19 @@ mod test_wasm_val_type {
         assert_eq!(f32::WASM_TYPE, ValType::F32);
         assert_eq!(f64::WASM_TYPE, ValType::F64);
         assert_eq!(i128::WASM_TYPE, ValType::V128);
-        assert_eq!(FuncRef::WASM_TYPE, ValType::FuncRef);
+        // assert_eq!(FuncRef::WASM_TYPE, ValType::FuncRef);
```
Comment (apepkuss) [inline src/io.rs:60]: Please remove it if it's not useful. Thanks!

File: src/io.rs [hunk 5]
```diff
@@ -208,21 +208,21 @@ mod test_wasm_val_type_list {
...
+        //         ValType::F32,
+        //         ValType::F64,
+        //         ValType::FuncRef,
+        //         ValType::ExternRef,
+        //     ]
+        // );
+        // assert_eq!(
+        //     <(FuncRef, ExternRef)>::wasm_types(),
+        //     [ValType::FuncRef, ValType::ExternRef]
+        // );
```
Comment (apepkuss) [inline src/io.rs:225]: Please remove them if they're not useful. Thanks!

File: .github/workflows/ci-build.yml (hunk 1 above)
Comment (L-jasmine) [reply .github/workflows/ci-build.yml:162]: Because Fedora cannot install the rustls plugin.

Comment (L-jasmine) [reply .github/workflows/ci-build.yml:162]: Because Fedora cannot install the rustls plugin.

**Comment by L-jasmine** [comment]
Because FuncRef typically has state, its lifecycle depends on the Module it resides in. Currently, the need to use FuncRef is mostly in the context of wasm-gc, and Rust compilation to wasm doesn't yet support the usage of FuncRef. Therefore, I've commented it out. However, since wasm-gc is nearing completion, we will need to consider how to safely export FuncRef when it's implemented. Hence, I believe there's no need to delete the current comments related to FuncRef.

**Pull request merged**